src/libstd/sys/wasm/thread_local_atomics.rs

```diff
@@ -8,22 +8,57 @@
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
 
+use sys::thread;
+use sync::atomic::{AtomicUsize, Ordering::SeqCst};
+
+const MAX_KEYS: usize = 128;
+static NEXT_KEY: AtomicUsize = AtomicUsize::new(0);
+
+struct ThreadControlBlock {
+    keys: [*mut u8; MAX_KEYS],
+}
+
+impl ThreadControlBlock {
+    fn new() -> ThreadControlBlock {
+        ThreadControlBlock {
+            keys: [0 as *mut u8; MAX_KEYS],
+        }
+    }
+
+    fn get() -> *mut ThreadControlBlock {
+        let ptr = thread::tcb_get();
+        if !ptr.is_null() {
+            return ptr as *mut ThreadControlBlock
+        }
+        let tcb = Box::into_raw(Box::new(ThreadControlBlock::new()));
+        thread::tcb_set(tcb as *mut u8);
+        tcb
+    }
+}
+
 pub type Key = usize;
 
-pub unsafe fn create(_dtor: Option<unsafe extern fn(*mut u8)>) -> Key {
-    panic!("TLS on wasm with atomics not implemented yet");
+pub unsafe fn create(dtor: Option<unsafe extern fn(*mut u8)>) -> Key {
+    drop(dtor); // FIXME: need to figure out how to hook thread exit to run this
+    let key = NEXT_KEY.fetch_add(1, SeqCst);
+    if key >= MAX_KEYS {
+        panic!("cannot allocate space for more TLS keys");
+    }
+    return key + 1 // offset by 1 so we never hand out 0
```

Comment:
Why don't we want to hand out 0?

Reply:
Ah this is a bit cryptic (and perhaps backwards), but it's due to [this block](https://github.com/rust-lang/rust/blob/e1041c6cd1a027ab3ada3e8538620d2e1d7067fe/src/libstd/sys_common/thread_local.rs#L169-L172) where out platform-agnostic thread local code can't deal with a TLS descriptor of 0 so it just asks for another one if that happens